### ISSUE TITLE
Tighten matches for roman numeral page numbers

### ISSUE THREAD

**Issue by jcushman** [freelawproject/eyecite#56]
Once I get eyecite running on the CAP corpus, I'd like to look at all of the extracted citations where the page number was supposedly a roman numeral, and figure out what the error rate is and if we can filter out common false matches or restrict roman numeral matches to some reporters and volumes.

Here are false positive matches I've collected so far:

- **49 Col. L.** Rev. 875
- **65 Mich. L.** Rev. 477
- **37 Taylor v.** Bd. of Educ., 240 Fed. Appx. 717 (6th Cir. 2007)
- **11 U. S. C.** § 701

(Moving this conversation over from #18 to focus specifically on roman numerals.)

**Comment by jcushman** [comment]
OK, I now have the data for running eyecite on CAP. I think the upshot is we (CAP at least) don't want to look for roman numeral page numbers at all, unless they match a list of known volumes with cases indexed by roman numeral. Here's the details ...

CAP has roughly 100 million full or short cites, of which roughly 500,000 have roman numeral page numbers. About 95% of those seem to be false positives.

Here are the ten most common roman numeral cites:

total | 501671
-- | --
28 U. S. C | 13879
42 U. S. C | 9941
18 U. S. C | 7975
42 Pa. C | 7191
29 U. S. C | 6345
28 U.S. C | 5668
15 U. S. C | 5333
6 R. C | 4881
26 U. S. C | 4009

All data: [roman_numerals_counts.csv](https://github.com/freelawproject/eyecite/files/6327154/roman_numerals_counts.csv)

Those are all false positives for "C", but looking down the list there's a lot of other common false positives as well -- "10 A. M", "7 H. L", "32 Hen. VIII" etc. (That last one is awesome! We're finding cites to the law of King Henry VIII.)

Here are counts for the top page numbers:

total | 501656
-- | --
C | 279821
L | 104392
l | 35970
M | 34795
D | 9512
v | 5608
I | 3740
m | 1180
VIII | 1163
x | 1077
vii | 957
xi | 911

All data: [roman_numerals_pages.csv](https://github.com/freelawproject/eyecite/files/6327166/roman_numerals_pages.csv)

I think these are basically all false positives down to `x` or `vii` or so. If we assume that all cites are false positives where `page in ['C', 'L', 'M', 'D', 'v', 'l', 'm', 'did']` or `reporter in ['Edw.', 'Hen.', 'Rich.', 'Will.', 'H.']` (the English kings), we end up with this:

total | 25332
-- | --
142 S. W. xii | 469
149 S. W. x | 435
142 S. W. xiii | 270
175 Ala. xxi | 226
198 Fed. xxvi | 167
38 Okla. vi | 157
137 Pac. ix | 156
115 C. C. A. xxvi | 152
117 Pac. xi | 150
89 Fed. vii | 134

All data: 
[roman_numerals_filtered.csv](https://github.com/freelawproject/eyecite/files/6327205/roman_numerals_filtered.csv)

These are actually looking reasonable! But we're now down to ~5% of the matches being likely real -- and many of those aren't going to be to cases at all, so even if I find the cite I can't show you anything for it.

To try to find exceptions to that rule I grouped "real" matches by volume, resulting in this:

total | 25332
-- | --
142 S. W. | 1366
198 Fed. | 918
115 C. C. A. | 843
89 Fed. | 687
32 C. C. A. | 567
149 S. W. | 438
175 Ala. | 308
169 S. W. | 287
38 Okla. | 276
150 Fed. | 270

All data: 
[roman_numerals_filtered_volumes.csv](https://github.com/freelawproject/eyecite/files/6327206/roman_numerals_filtered_volumes.csv)

So roughly a quarter of "real" cites point to those ten volumes, which might mean they really have cases in them. But you'd still need a few hundred exceptions to cover everything.

The other option I considered is to tighten the roman numerals regex to only accept numerals up to 200. Something simplified that amounts to this:

I|II|III|IV|V|VI|VII|VIII|IX|X|XI|XII|XIII|XIV|XV|XVI|XVII|XVIII|XIX|XX|XXI|XXII|XXIII|XXIV|XXV|XXVI|XXVII|XXVIII|XXIX|XXX|XXXI|XXXII|XXXIII|XXXIV|XXXV|XXXVI|XXXVII|XXXVIII|XXXIX|XL|XLI|XLII|XLIII|XLIV|XLV|XLVI|XLVII|XLVIII|XLIX|L|LI|LII|LIII|LIV|LV|LVI|LVII|LVIII|LIX|LX|LXI|LXII|LXIII|LXIV|LXV|LXVI|LXVII|LXVIII|LXIX|LXX|LXXI|LXXII|LXXIII|LXXIV|LXXV|LXXVI|LXXVII|LXXVIII|LXXIX|LXXX|LXXXI|LXXXII|LXXXIII|LXXXIV|LXXXV|LXXXVI|LXXXVII|LXXXVIII|LXXXIX|XC|XCI|XCII|XCIII|XCIV|XCV|XCVI|XCVII|XCVIII|XCIX|C|CI|CII|CIII|CIV|CV|CVI|CVII|CVIII|CIX|CX|CXI|CXII|CXIII|CXIV|CXV|CXVI|CXVII|CXVIII|CXIX|CXX|CXXI|CXXII|CXXIII|CXXIV|CXXV|CXXVI|CXXVII|CXXVIII|CXXIX|CXXX|CXXXI|CXXXII|CXXXIII|CXXXIV|CXXXV|CXXXVI|CXXXVII|CXXXVIII|CXXXIX|CXL|CXLI|CXLII|CXLIII|CXLIV|CXLV|CXLVI|CXLVII|CXLVIII|CXLIX|CL|CLI|CLII|CLIII|CLIV|CLV|CLVI|CLVII|CLVIII|CLIX|CLX|CLXI|CLXII|CLXIII|CLXIV|CLXV|CLXVI|CLXVII|CLXVIII|CLXIX|CLXX|CLXXI|CLXXII|CLXXIII|CLXXIV|CLXXV|CLXXVI|CLXXVII|CLXXVIII|CLXXIX|CLXXX|CLXXXI|CLXXXII|CLXXXIII|CLXXXIV|CLXXXV|CLXXXVI|CLXXXVII|CLXXXVIII|CLXXXIX|CXC|CXCI|CXCII|CXCIII|CXCIV|CXCV|CXCVI|CXCVII|CXCVIII|CXCIX|CC

Which would be a lot better, but `C`, `L`, `l`, `I`, `v`, and maybe `x` would still have to be excluded. So you end up with a weird sort of API where eyecite is like "we extract roman numeral cites up to 200, except for cites to page 1, 5, 10, 50, or 100 which we ignore."

So, for now CAP is just not looking for these at all, by setting `eyecite.utils.PAGE_NUMBER_REGEX = r"\d+"`. I'm curious what looks like the best compromise to other folks though.

**Comment by mlissner** [comment]
I foolishly put this in Slack, but it looks like the roman numerals are coming from freelawproject/courtlistener#510, which was to address some roman numerals we found in citations:

```
218 Neb. xxiii (1984)
250 Neb. xxiv (1996)
272 Neb. xxxiii (2007)
269 Neb. xxvi (2004)
14 Neb. App. liv (2006)
12 Neb. App. lxiv (2004)
13 Neb. App. liii (2005)
13 Neb. App. xvi (2005)
14 Neb. App. lxi (2006)
13 Neb. App. l (2005)
14 Neb. App. xxvii (2006)
267 Ga. XXVI (1996)
367 Ark. xx (2006)
```

Most of those, I note, are lowercase, which might help with this a bit. Back in 2016, I confirmed with a librarian in Nebraska that these were citations to the front matter of the volume. And indeed, we do have cases at those locations somehow:

[0c5c0dadba8375ea.txt](https://github.com/freelawproject/eyecite/files/6328635/0c5c0dadba8375ea.txt)

**Comment by mlissner** [comment]
Reading and re-reading your notes and CSVs, I kind of feel like limiting ourselves to roman numerals less than 100 that are lowercase might eliminate almost all the false positives. I think making a blocklist of non-reporters might push it into being actually useful.

**Comment by jcushman** [comment]
Those do look like the best ones. Here's the counts for just those in numerical order (e.g, `v` appears 5608 times):

total | 58776 |  
-- | -- | --
1 | 241 | i
2 | 23 | ii
3 | 73 | iii
4 | 130 | iv
5 | 5608 | v
6 | 701 | vi
7 | 957 | vii
8 | 807 | viii
9 | 788 | ix

[roman_numerals_lower_200.csv](https://github.com/freelawproject/eyecite/files/6329989/roman_numerals_lower_200.csv)

The one issue with this is `v`, `l`, and `c` are clearly mostly false positives, and `i` and `x` are a little suspiciously high as well. So about 2/3 are still false positive -- if you take out `v`, `l`, and `c` entirely you end up with 16,982 valid-ish matches.

I'm not sure if a blocklist ends up helping to revive the `v` and `l` cites -- something like "Mass. v" is viable as a cite or a case name, and "Mass. l." is viable as a cite or as "Mass. laws". Here's all the reporter-pages for `ivxlc` if you want to see those: [roman_numerals_common_pages.csv](https://github.com/freelawproject/eyecite/files/6329996/roman_numerals_common_pages.csv)

I think it ends up being a question of whether it's better to grab all the low-number lowercase except for `vlc`, or to opt-in specific reporters like Neb. and Neb. App. that have cases at roman numerals.

**Comment by mlissner** [comment]
> I think it ends up being a question of whether it's better to grab all the low-number lowercase except for vlc, or to opt-in specific reporters like Neb. and Neb. App. that have cases at roman numerals.

Yeah, I agree. I don't know all of the reporters that we'd want to opt in, and we'd never be done updating that list as new reporters came out. I think I lean towards "all roman minus `vlc`." 

As a user, that feels like a mostly fine compromise. You wouldn't see too many false positives, and yeah, we'll miss a few opportunities to link things here and there.

**Comment by mlissner** [comment]
Fixed by #61
